code/modules/wiremod/shell/brain_computer_interface.dm

```diff
@@ -28,62 +31,41 @@
 	else
 		return ..()
 
-/obj/item/circuit_component/equipment_action/bci
-	display_name = "BCI Action"
-	desc = "Represents an action the user can take when implanted with the brain-computer interface."
-	required_shells = list(/obj/item/organ/internal/cyberimp/bci)
-
-	/// A reference to the action button itself
-	var/datum/action/innate/bci_action/bci_action
-
-/obj/item/circuit_component/equipment_action/bci/Destroy()
-	QDEL_NULL(bci_action)
-	return ..()
-
-/obj/item/circuit_component/equipment_action/bci/register_shell(atom/movable/shell)
-	. = ..()
-	var/obj/item/organ/internal/cyberimp/bci/bci = shell
-	if(istype(bci))
-		bci_action = new(src)
-		update_action()
-
-		bci.actions += list(bci_action)
-
-/obj/item/circuit_component/equipment_action/bci/unregister_shell(atom/movable/shell)
-	var/obj/item/organ/internal/cyberimp/bci/bci = shell
-	if(istype(bci))
-		bci.actions -= bci_action
-		QDEL_NULL(bci_action)
-	return ..()
-
-/obj/item/circuit_component/equipment_action/bci/input_received(datum/port/input/port)
-	if (!isnull(bci_action))
-		update_action()
+/obj/item/organ/internal/cyberimp/bci/proc/action_comp_registered(datum/source, obj/item/circuit_component/equipment_action/action_comp)
+	SIGNAL_HANDLER
+	LAZYADD(actions, new/datum/action/innate/bci_action(src, action_comp))
 
-/obj/item/circuit_component/equipment_action/bci/update_action()
-	bci_action.name = button_name.value
-	bci_action.button_icon_state = "bci_[replacetextEx(lowertext(icon_options.value), " ", "_")]"
+/obj/item/organ/internal/cyberimp/bci/proc/action_comp_unregistered(datum/source, obj/item/circuit_component/equipment_action/action_comp)
+	SIGNAL_HANDLER
+	var/datum/action/innate/bci_action/action = action_comp.granted_to[REF(src)]
+	if(!istype(action))
+		return
+	LAZYREMOVE(actions, action)
+	qdel(action)
 
 /datum/action/innate/bci_action
 	name = "Action"
 	button_icon = 'icons/mob/actions/actions_items.dmi'
 	check_flags = AB_CHECK_CONSCIOUS
 	button_icon_state = "bci_power"
 
-	var/obj/item/circuit_component/equipment_action/bci/circuit_component
+	var/obj/item/circuit_component/equipment_action/circuit_component
 
-/datum/action/innate/bci_action/New(obj/item/circuit_component/equipment_action/bci/circuit_component)
+/datum/action/innate/bci_action/New(obj/item/organ/internal/cyberimp/bci/bci, obj/item/circuit_component/equipment_action/circuit_component)
 	..()
-
+	circuit_component.granted_to[REF(bci)] = src
 	src.circuit_component = circuit_component
 
 /datum/action/innate/bci_action/Destroy()
-	circuit_component.bci_action = null
+	for(var/ref in circuit_component.granted_to)
+		if (circuit_component.granted_to[ref] == src)
+			circuit_component.granted_to -= ref
```

Comment:
it would make more sense to store a ref to the BCI and clear it that way instead of iterating through this list.